### PR TITLE
add safeCast operator

### PR DESCRIPTION
**Pull request by hoangchungk53qx1**
## Status

**READY**

## Breaking Changes

NO

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [x] 📝 Documentation
- [x] 🗑️ Chore

## Requirements

- [ ✅ ] Dump api
- [ ✅ ] Update README
- [ ✅ ] Update CHANGELOG
- [ ✅ ] Add tests
- [ ✅ ] Add docs

### CODE REVIEW

File: src/commonTest/kotlin/com/hoc081098/flowext/CastTest.kt [hunk 1]
```diff
@@ -96,4 +96,33 @@ class CastTest : BaseTest() {
     val flow = flowOf(1, 2, 3)
     assertSame(flow.castNullable(), flow)
   }
+
+  @Test
+  fun testSafeCastSuccess() = runTest {
+    assertIs<Flow<Int>>(
```
Comment (hoc081098) [inline src/commonTest/kotlin/com/hoc081098/flowext/CastTest.kt:102]:
```suggestion
    assertIs<Flow<Int?>>(
```

File: src/commonTest/kotlin/com/hoc081098/flowext/CastTest.kt [hunk 2]
```diff
@@ -96,4 +96,33 @@ class CastTest : BaseTest() {
...
+      flowOf<Any?>(1, 2, 3, "hello").safeCast<Int?>(),
+    )
+
+    assertIs<Flow<Int>>(
+      flowOf<Int?>(1, 2, 3)
```
Comment (hoc081098) [inline src/commonTest/kotlin/com/hoc081098/flowext/CastTest.kt:107]:
```suggestion
      flowOf(1, 2, 3)
```

File: src/commonTest/kotlin/com/hoc081098/flowext/CastTest.kt [hunk 3]
```diff
@@ -96,4 +96,33 @@ class CastTest : BaseTest() {
...
     assertSame(flow.castNullable(), flow)
   }
+
+  @Test
+  fun testSafeCastSuccess() = runTest {
+    assertIs<Flow<Int>>(
+      flowOf<Any?>(1, 2, 3, "hello").safeCast<Int?>(),
+    )
+
+    assertIs<Flow<Int>>(
```
Comment (hoc081098) [inline src/commonTest/kotlin/com/hoc081098/flowext/CastTest.kt:106]:
```suggestion
    assertIs<Flow<Int?>>(
```

File: src/commonTest/kotlin/com/hoc081098/flowext/CastTest.kt [hunk 4]
```diff
@@ -96,4 +96,33 @@ class CastTest : BaseTest() {
...
+      )
+  }
+
+  @Test
+  fun testSafeCast() {
+    val stringFlow: Flow<String?> = flowOf("Hello", 42, "World", 123, "Kotlin").safeCast()
+    assertIs<Flow<String?>>(stringFlow)
+
+    val intFlow: Flow<Int?> = flowOf(1, 2, 3, null).safeCast()
+    assertIs<Flow<Int?>>(intFlow)
```
Comment (hoc081098) [inline src/commonTest/kotlin/com/hoc081098/flowext/CastTest.kt:126]:
```suggestion
    val stringFlow: Flow<String?> = flowOf("Hello", 42, "World", 123, "Kotlin").safeCast()
    assertIs<Flow<String?>>(stringFlow)
    assertContentEquals(
      listOf(...),
      stringFlow.toList(),
    )

    val intFlow: Flow<Int?> = flowOf(1, 2, 3, null).safeCast()
    assertIs<Flow<Int?>>(intFlow)
    assertContentEquals(
      listOf(...),
      intFlow.toList(),
    )
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,10 @@
 
 ## [Unreleased] - TODO
 
+### Added
+
+- Add `Flow.safeCast` operator.
+
```
Comment (hoc081098) [inline CHANGELOG.md:8]: `Added section` is below

File: src/commonMain/kotlin/com/hoc081098/flowext/cast.kt [hunk 6]
```diff
@@ -53,3 +53,9 @@ public inline fun <reified T : Any> Flow<T?>.castNotNull(): Flow<T> = map { it!!
  */
 @Suppress("NOTHING_TO_INLINE")
 public inline fun <T> Flow<T>.castNullable(): Flow<T?> = this
+
+/**
+ * Adapt this `Flow<*>` to be a `Flow<R?>`.
+ */
```
Comment (hoc081098) [inline src/commonMain/kotlin/com/hoc081098/flowext/cast.kt:59]: Adapt this `Flow<*>` to be a `Flow<R?>`.
At the collection time, if this [Flow] has any value that is not an instance of R, `null` will be emitted

File: README.md [hunk 7]
```diff
@@ -487,6 +488,37 @@ castNotNull: 3
 
 ----
 
+
+##### safeCast
+
+Adapt this `Flow<*>` to be a `Flow<R?>`.
+
+At the collection time, converts a Flow<*> to a Flow<R?>, if not consistent with R, returns null.
```
Comment (hoc081098) [inline README.md:496]: Same as docs in kotlin 🙏
> Adapt this `Flow<*>` to be a `Flow<R?>`.
At the collection time, if this Flow has any value that is not an instance of R, `null` will be emitted

File: README.md [hunk 8]
```diff
@@ -487,6 +488,37 @@ castNotNull: 3
...
+```kotlin
+flowOf<Any?>(1, 2, 3, "Kotlin", null)
+  .safeCast<Int?>()
+  .collect { v: Int? ->
+    if (v != null) {
+      println("safeCast: $v")
+    } else {
+      println("Null value encountered.")
+    }
+  }
```
Comment (hoc081098) [inline README.md:507]:
```suggestion
  .collect { v: Int? -> println("safeCast: $v") }
```

File: README.md [hunk 9]
```diff
@@ -487,6 +488,37 @@ castNotNull: 3
...
+```
+
+Output:
+
+```none
+safeCast: 1
+safeCast: 2
+safeCast: 3
+Null value encountered.
+Null value encountered.
```
Comment (hoc081098) [inline README.md:517]: https://github.com/hoc081098/FlowExt/pull/181/files#r1348959726

**Review by hoc081098 — COMMENTED** [review]
🙏🙏🙏

**Comment by hoangchungk53qx1** [comment]
resolve comment @hoc081098  🙏 🥇 💯 🙏

File: CHANGELOG.md [hunk 10]
```diff
@@ -4,6 +4,10 @@
 
 ### Changed
 
+### Added
+
+- Add `Flow.safeCast` operator.
```
Comment (hoc081098) [inline CHANGELOG.md:9]: Move to line 18 (after ignoreElements, scanWith)

File: src/commonMain/kotlin/com/hoc081098/flowext/cast.kt [hunk 11]
```diff
@@ -53,3 +53,9 @@ public inline fun <reified T : Any> Flow<T?>.castNotNull(): Flow<T> = map { it!!
  */
 @Suppress("NOTHING_TO_INLINE")
 public inline fun <T> Flow<T>.castNullable(): Flow<T?> = this
+
+/**
+ * Adapt this `Flow<*>` to be a `Flow<R?>`.
```
Comment (hoc081098) [inline src/commonMain/kotlin/com/hoc081098/flowext/cast.kt:58]: Add `At the collection time, if this [Flow] has any value that is not an instance of R, null will be emitted`

File: src/commonTest/kotlin/com/hoc081098/flowext/CastTest.kt [hunk 12]
```diff
@@ -96,4 +98,48 @@ class CastTest : BaseTest() {
...
+
+    kotlinx.coroutines.test.runTest {
+      assertContentEquals(
+        listOf(1, 2, 3, null),
+        intFlow.toList(),
+      )
+    }
+  }
```
Comment (hoc081098) [inline src/commonTest/kotlin/com/hoc081098/flowext/CastTest.kt:144]:
```suggestion
  fun testSafeCast() = runTest {
    val stringFlow: Flow<String?> = flowOf("Hello", 42, "World", 123, "Kotlin").safeCast()
    assertIs<Flow<String?>>(stringFlow)

   
    assertContentEquals(
      listOf("Hello", null, "World", null, "Kotlin"),
      stringFlow.toList(),
    )
   
    val intFlow: Flow<Int?> = flowOf(1, 2, 3, null).safeCast()
    assertIs<Flow<Int?>>(intFlow)

    assertContentEquals(
      listOf(1, 2, 3, null),
      intFlow.toList(),
    )
  }
```

**Review by hoc081098 — COMMENTED** [review]
🙏❤😁

**Comment by hoangchungk53qx1** [comment]
@hoc081098 everything done

File: src/commonMain/kotlin/com/hoc081098/flowext/cast.kt [hunk 13]
```diff
@@ -53,3 +53,11 @@ public inline fun <reified T : Any> Flow<T?>.castNotNull(): Flow<T> = map { it!!
...
+ * Add At the collection time, if this [Flow] has any value that is not an instance of R,
```
Comment (hoc081098) [inline src/commonMain/kotlin/com/hoc081098/flowext/cast.kt:59]: @hoangchungk53qx1 dư chữ `Add  `🤣🤣

**Review by hoc081098 — APPROVED** [review]
Xincamon ❤🙆🏩

File: src/commonMain/kotlin/com/hoc081098/flowext/cast.kt (hunk 13 above)
Comment (hoangchungk53qx1) [reply src/commonMain/kotlin/com/hoc081098/flowext/cast.kt]: a serious error 🍔

**Pull request merged**